format/macho/macho.go

```diff
@@ -0,0 +1,835 @@
+				d.FieldU32("id")
+			})
+		case LC_IDFVMLIB, LC_LOADFVMLIB:
+			d.FieldStruct("fvmlib", func(d *decode.D) {
+				offset := d.FieldU32("offset")
+				d.FieldU32("minor_version")
+				d.FieldU32("header_addr", scalar.Hex)
+				d.FieldUTF8NullFixedLen("name", int(cmdsize)-int(offset))
+			})
+		default:
+			if _, ok := loadCommands[cmd]; !ok {
+				d.FieldRawLen("unknown", int64((cmdsize-8)*8))
+			}
+		}
+		ncmdsIdx++
+	})
+}
+
+func fatParse(d *decode.D) {
+	// Go to start of the file again
+	d.SeekAbs(0)
+	d.FieldStruct("fat_header", func(d *decode.D) {
+		d.FieldRawLen("magic", 8*8)
+		narchs := d.FieldU32("narchs")
+		narchsIdx := 0
+		d.FieldStructArrayLoop("archs", "arch", func() bool {
+			return narchsIdx < int(narchs)
+		}, func(d *decode.D) {
+			// parse FatArch
+			d.FieldStruct("fat_arch", func(d *decode.D) {
+				// beware cputype and cpusubtype changes from ofile header to fat header
+				cpuType := d.FieldU32("cputype", cpuTypes)
+				d.FieldU32("cpusubtype", cpuSubTypes[int64(cpuType)])
+				d.FieldU32("offset")
+				d.FieldU32("size")
+				d.FieldU32("align")
+			})
+		})
+		for i := 0; uint64(i) < narchs; i++ {
+			// parse ofiles
+			ofileDecode(d)
+		}
+	})
+}
+
+func intelSubTypeHelper(f, m int64) int64 {
+	return f + (m << 4)
+}
+
+func parseFlags(symbolMap map[uint64]string) func(*decode.D) {
+	return func(d *decode.D) {
+		var flagIdx uint64
+		flags := d.FieldU32("flags")
+		for flagIdx = 1; flagIdx <= 0x80000000; flagIdx <<= 1 {
+			if val, ok := symbolMap[flagIdx]; ok {
+				d.FieldValueBool(val, (flagIdx&flags) != 0)
+			}
+		}
+	}
+}
+
+var timestampMapper = scalar.Fn(func(s scalar.S) (scalar.S, error) {
+	ts, ok := s.Actual.(uint64)
+	if !ok {
+		return s, nil
+	}
+	s.Sym = time.UnixMilli(int64(ts)).UTC().String()
+	return s, nil
+})
+
+func threadStateI386Decode(d *decode.D) {
+	d.FieldU32("eax")
+	d.FieldU32("ebx")
+	d.FieldU32("ecx")
+	d.FieldU32("edx")
+	d.FieldU32("edi")
+	d.FieldU32("esi")
+	d.FieldU32("ebp")
+	d.FieldU32("esp")
+	d.FieldU32("ss")
+	d.FieldU32("eflags")
+	d.FieldU32("eip")
+	d.FieldU32("cs")
+	d.FieldU32("ds")
+	d.FieldU32("es")
+	d.FieldU32("fs")
+	d.FieldU32("gs")
+}
+
+func threadStateX8664Decode(d *decode.D) {
+	d.FieldU64("rax")
+	d.FieldU64("rbx")
+	d.FieldU64("rcx")
+	d.FieldU64("rdx")
+	d.FieldU64("rdi")
+	d.FieldU64("rsi")
+	d.FieldU64("rbp")
+	d.FieldU64("rsp")
+	d.FieldU64("r8")
+	d.FieldU64("r9")
+	d.FieldU64("r10")
+	d.FieldU64("r11")
+	d.FieldU64("r12")
+	d.FieldU64("r13")
+	d.FieldU64("r14")
+	d.FieldU64("r15")
+	d.FieldU64("rip")
+	d.FieldU64("rflags")
+	d.FieldU64("cs")
+	d.FieldU64("fs")
+	d.FieldU64("gs")
+}
+
+func threadStateARM32Decode(d *decode.D) {
+	rIdx := 0
+	d.FieldStructArrayLoop("r", "r", func() bool {
+		return rIdx < 13
+	}, func(d *decode.D) {
+		d.FieldU32("value")
+	})
+	d.FieldU32("sp")
+	d.FieldU32("lr")
+	d.FieldU32("pc")
+	d.FieldU32("cpsr")
+}
+
+func threadStateARM64Decode(d *decode.D) {
+	rIdx := 0
+	d.FieldStructArrayLoop("r", "r", func() bool {
```

Comment:
Some of the FieldStructArrayLoop seems to be missing increment of the loop variable?